### PR TITLE
Run timeout-based tests only in CI

### PR DESCRIPTION
**Pull request by s0me0ne-unkn0wn**
Integration tests checking PVF preparation and execution queues behavior are known to fail sometimes when run locally by developers on heavily loaded machines. Still, they are crucial and cannot be easily replaced.

This PR aims to run those tests in CI environment only, putting them behind a dedicated feature and running the CI tests with the corresponding feature flag.

Related to: paritytech/substrate#11321

### CODE REVIEW

**Comment by bkchr** [comment]
When these tests fail locally, they can also fail on CI when there is high load. This pr will not fix this.

**Comment by s0me0ne-unkn0wn** [comment]
They can, but
1) They never did (or it has never been reported);
2) We still need them;
3) As far as I can see, rewriting them as non-timeout-based is a huge complication that doesn't pay off.

This PR makes the situation better, not perfect. Here we declare it's okay for a timeout-based test to fail on someone's laptop which is running 128 instances of `rust-analyzer` at the same time. It is not okay to fail in CI though. If it starts failing in CI, then the test should be fixed somehow.

Any ideas on how to improve the tests themselves are welcome, too.

**Review by ggwpez — COMMENTED** [review]
Why not use an env-var? This will fail with `--all-features`.

**Comment by s0me0ne-unkn0wn** [comment]
> Why not use an env-var?

Do you have an example of how to skip a test at compile time using env-var? All the solutions coming to mind involve some checking at runtime which would make things much less explicit.

> This will fail with `--all-features`.

Well, I'd put it differently: with `--all-features` the user requests to test all features, including tests that are known to be flaky in local environments, and that MAY fail. It's explicit enough.

But I don't care much because testing with `--all-features` is broken in Polkadot anyways. `cargo +stable test --profile testnet --all --all-features` doesn't even build.

File: node/core/pvf/Cargo.toml [hunk 1]
```diff
@@ -45,3 +45,6 @@ hex-literal = "0.3.4"
 
 adder = { package = "test-parachain-adder", path = "../../../parachain/test-parachains/adder" }
 halt = { package = "test-parachain-halt", path = "../../../parachain/test-parachains/halt" }
+
+[features]
```
Comment (sandreim) [inline node/core/pvf/Cargo.toml:49]: Would be great to drop in a comment here to document the usage of the feature

**Review by sandreim — APPROVED** [review]
I like it better than ENV vars

File: Cargo.toml [hunk 2]
```diff
@@ -210,6 +211,10 @@ runtime-metrics = [ "polkadot-cli/runtime-metrics" ]
 pyroscope = ["polkadot-cli/pyroscope"]
 jemalloc-allocator = ["polkadot-node-core-pvf-prepare-worker/jemalloc-allocator", "polkadot-overseer/jemalloc-allocator"]
 
+# Enables timeout-based tests supposed to be run only in CI environment as they may be flaky
+# when run locally depending on system load
```
Comment (bkchr) [inline Cargo.toml:215]: This argument is just plainly wrong. CI systems are most of the time under high load and are most of the time systems where you see flaky behavior.

Comment (s0me0ne-unkn0wn) [reply Cargo.toml:215]: In theory, you're most probably right. In practice, we've never seen them being flaky in CI, but Gav has seen them being flaky locally. Remember, it was your mantra, "let's solve the problems we have, not the problems we don't have". That's what I'm trying to do, to describe and fix the real-world problem we're experiencing. If we ever see them fail in CI, the description and the fix will change.

Comment (bkchr) [reply Cargo.toml:215]: My mantra is to never use time based waiting in CI tests ;) As I have seen every combination of this failing in CI.

**Review by altaua — APPROVED** [review]

File: Cargo.toml (hunk 2 above)
Comment (altaua) [reply Cargo.toml:215]: We currently dedicate a vm to each compute-intensive ci job run, so high load situations interfering with tests should indeed be somewhat less likely in our ci compared to local runs.

Comment (bkchr) [reply Cargo.toml:215]: Okay. Good to know. I mean this wasn't always the case, but fine if that changed.

**Comment by s0me0ne-unkn0wn** [comment]
bot merge

**Pull request merged**